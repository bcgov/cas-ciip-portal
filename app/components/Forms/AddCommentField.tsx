import React, {useState} from 'react';
import {FieldProps} from '@rjsf/core';
import {Form, Button} from 'react-bootstrap';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {faCommentAlt} from '@fortawesome/free-solid-svg-icons';
import {toast} from 'react-toastify';

const AddCommentField: React.FunctionComponent<FieldProps<string>> = (
  props
) => {
  const [expanded, setExpanded] = useState(false);
  const [value, setValue] = useState(props.formData);
  const title = props.formData ? 'Edit comment' : 'Add a comment';

  const handleChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    setValue(e.target.value);
  };

  const handleCancel = () => {
    setValue(props.formData);
    setExpanded(false);
  };

  const handleSave = () => {
    props.onChange(value);
    setExpanded(false);
    toast.success(
      'Your comment was saved for review upon application submission',
      {
        autoClose: 5000,
        position: 'top-center'
      }
    );
  };

  return (
    <>
      <style jsx>{`
        .title,
        .report-field .controls {
          display: flex;
          width: 100%;
          justify-content: flex-end;
          align-items: center;
        }

        .report-field .controls {
          margin-top: 16px;
        }

        .report-field {
          background-color: #fcf8e2;
          border: 1px solid #ccc;
          border-radius: 4px;
          padding: 16px;
          margin-top: 16px;
        }
      `}</style>
      <span className="title">
        <Button variant="link" onClick={() => setExpanded(true)}>
          <FontAwesomeIcon icon={faCommentAlt} />
          &nbsp;
          {title}
        </Button>
      </span>
      {expanded && (
        <div className="report-field">
          <legend>CIIP Application Comment</legend>

          <p>
            This area is only for leaving a comment regarding the data that you
            reported on this page. This comment will be visible by the person
            reviewing this CIIP application once the application is submitted.
          </p>
          <p>
            For any technical issue, or to review documentation, please click on
            the help bubble in the bottom right of your screen.
          </p>
          <Form.Control
            as="textarea"
            rows={8}
            value={value}
            onChange={handleChange}
          />
          <div className="controls">
            <Button variant="outline-dark" onClick={handleCancel}>
              Cancel
            </Button>
            <Button
              style={{marginLeft: 16}}
              variant="success"
              onClick={handleSave}
            >
              Save Comment
            </Button>
          </div>
        </div>
      )}
    </>
  );
};

export default AddCommentField;
