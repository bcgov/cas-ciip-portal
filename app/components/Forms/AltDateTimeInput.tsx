// https://github.com/rjsf-team/react-jsonschema-form/blob/v1.8.1/src/components/widgets/AltDateTimeWidget.js

import React from "react";
import PropTypes from "prop-types";
import AltDateInput from "./AltDateInput";
import { WidgetProps } from "react-jsonschema-form";

function AltDateTimeInput(props: WidgetProps) {
  const { AltDateInput } = props.registry.widgets;
  return <AltDateInput time {...props} />;
}

if (process.env.NODE_ENV !== "production") {
  AltDateTimeInput.propTypes = {
    schema: PropTypes.object.isRequired,
    id: PropTypes.string.isRequired,
    value: PropTypes.string,
    required: PropTypes.bool,
    onChange: PropTypes.func,
    options: PropTypes.object,
  };
}

AltDateTimeInput.defaultProps = {
  ...AltDateInput.defaultProps,
  time: true,
};

export default AltDateTimeInput;