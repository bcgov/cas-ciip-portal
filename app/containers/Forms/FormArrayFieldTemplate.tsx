import React from 'react';
import {ArrayFieldTemplateProps} from 'react-jsonschema-form';
import {Form, Button, Col} from 'react-bootstrap';

const FormArrayFieldTemplate: React.FunctionComponent<ArrayFieldTemplateProps> = ({
  items,
  canAdd,
  onAddClick,
  uiSchema,
  className
}) => {
  return (
    <div className={className}>
      {items.map(element => {
        return (
          <React.Fragment key={element.index}>
            <Form.Row>
              <Col xs={12} md={element.hasRemove ? 10 : 12}>
                {element.children}
              </Col>

              {element.hasRemove && (
                <Col xs={12} md={2}>
                  <div className="remove-button-container">
                    <Button
                      type="button"
                      variant="danger"
                      onClick={element.onDropIndexClick(element.index)}
                    >
                      {uiSchema['ui:remove-text'] || 'Remove'}
                    </Button>
                  </div>
                </Col>
              )}
            </Form.Row>
            <hr />
          </React.Fragment>
        );
      })}
      {canAdd && (
        <Form.Row>
          <Col xs={12}>
            <div className="add-button-container">
              <Button type="button" onClick={onAddClick}>
                {uiSchema['ui:add-text'] || 'Add'}
              </Button>
            </div>
          </Col>
        </Form.Row>
      )}
      <style jsx>
        {`
          .remove-button-container {
            height: 100%;
            display: flex;
            justify-content: space-around;
            align-items: center;
          }
          .add-button-container {
            height: 100%;
            display: flex;
            justify-content: space-around;
            align-items: center;
          }
        `}
      </style>
    </div>
  );
};

export default FormArrayFieldTemplate;
