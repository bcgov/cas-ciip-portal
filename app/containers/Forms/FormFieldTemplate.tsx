import React from 'react';
import {Col, Form} from 'react-bootstrap';
import {FieldTemplateProps} from 'react-jsonschema-form';

const FormFieldTemplate: React.FunctionComponent<FieldTemplateProps> = ({
  id,
  label,
  help,
  required,
  description,
  errors,
  children,
  schema,
  classNames,
  uiSchema,
  hidden
}) => {
  if (hidden) return <div className="hidden">{children}</div>;

  const helpText = uiSchema['ui:helplink'] ? (
    <a
      className="font-italic"
      href={uiSchema['ui:helplink']}
      target="_blank"
      rel="noopener noreferrer"
    >
      {help}
    </a>
  ) : (
    <span className="font-italic text-muted">{help}</span>
  );

  if (schema.type === 'array')
    return (
      <Col xs={12}>
        {description}
        {helpText}
        {errors}
        {children}
      </Col>
    );
  if (schema.type === 'object')
    return (
      <Col xs={12}>
        <Form.Row>
          {description}
          {errors}
          {helpText}
          {children}
        </Form.Row>
      </Col>
    );

  const xs = uiSchema['ui:col-xs'] || 12;
  const sm = uiSchema['ui:col-sm'] || xs;
  const md = uiSchema['ui:col-md'] || (sm < 6 ? sm : 6);
  const lg = uiSchema['ui:col-lg'] || md;
  const xl = uiSchema['ui:col-xl'] || lg;

  return (
    <Form.Group
      as={Col}
      xs={xs}
      sm={sm}
      md={md}
      lg={lg}
      xl={xl}
      className={classNames}
    >
      <Form.Label htmlFor={id}>
        {label}
        {required ? <>&nbsp;*</> : null}
      </Form.Label>
      {description}
      {children}
      {errors}
      {helpText}
    </Form.Group>
  );
};

export default FormFieldTemplate;
