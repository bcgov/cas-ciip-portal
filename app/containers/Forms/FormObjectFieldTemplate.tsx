import React from 'react';
import {Col} from 'react-bootstrap';

const FormObjectFieldTemplate = ({
  TitleField,
  properties,
  title,
  description
}) => {
  return (
    <>
      {title && (
        <Col xs={12}>
          <TitleField title={title} />
        </Col>
      )}
      {properties.map(prop => (
        <React.Fragment key={prop.content.key}>{prop.content}</React.Fragment>
      ))}
      {description}
    </>
  );
};

export default FormObjectFieldTemplate;
