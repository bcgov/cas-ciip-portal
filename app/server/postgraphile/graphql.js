const {
  createPostGraphileSchema,
  withPostGraphileContext,
} = require("postgraphile");
const { graphql } = require("graphql");
const authenticationPgSettings = require("./authenticationPgSettings");
const { pgPool, getDatabaseUrl } = require("../db/databaseConnectionService");
const postgraphileOptions = require("./postgraphileOptions");

const pgSettings = (req) => {
  const opts = {
    ...authenticationPgSettings(req),
  };
  return opts;
};

let postgraphileSchemaSingleton;

const postgraphileSchema = async () => {
  if (!postgraphileSchemaSingleton) {
    postgraphileSchemaSingleton = await createPostGraphileSchema(
      getDatabaseUrl(),
      process.env.DATABASE_SCHEMA || "ggircs_portal",
      postgraphileOptions()
    );
  }

  return postgraphileSchemaSingleton;
};

async function performQuery(query, variables, request) {
  const settings = pgSettings(request);

  return withPostGraphileContext(
    {
      pgPool,
      pgSettings: settings,
    },
    async (context) => {
      // `context` object, which should NOT be used outside of this
      // function.
      const result = graphql(
        await postgraphileSchema(),
        query,
        null,
        { ...context }, // You can add more to context if you like
        variables
      );
      return result;
    }
  );
}

module.exports = {
  performQuery,
};
