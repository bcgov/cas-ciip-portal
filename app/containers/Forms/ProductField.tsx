import React from 'react';
import {Alert} from 'react-bootstrap';
import {FieldProps} from 'react-jsonschema-form';
import ObjectField from 'react-jsonschema-form/lib/components/fields/ObjectField';
import {createFragmentContainer, graphql} from 'react-relay';
import {ProductField_query} from 'ProductField_query.graphql';

interface FormData {
  productRowId?: number;
  productAmount?: number;
  productUnits?: string;
  productEmissions?: number;
  requiresEmissionAllocation?: boolean;
  requiresProductAmount?: boolean;
}

interface Props extends FieldProps<FormData> {
  query: ProductField_query;
}

/**
 * This custom ObjectField component injects the read-only data for a product when the
 * product id changes
 */
export const ProductFieldComponent: React.FunctionComponent<Props> = (
  props
) => {
  const {formData, query, onChange} = props;

  // TODO: Clean up this function with array.some() as noted in https://github.com/bcgov/cas-ciip-portal/pull/621
  const productIsPublished = (formData, query) => {
    const product = query.allProducts.edges.find(
      ({node}) => node.rowId === formData.productRowId
    )?.node;
    if (product?.productState === 'PUBLISHED' || !product) return true;
    return false;
  };

  const handleProductChange = (productRowId: number) => {
    const product = query.allProducts.edges.find(
      ({node}) => node.rowId === productRowId
    )?.node;
    onChange({
      ...formData,
      productRowId,
      productUnits: product?.units,
      requiresEmissionAllocation: product?.requiresEmissionAllocation,
      requiresProductAmount: product?.requiresProductAmount,
      isEnergyProduct: product?.isEnergyProduct,
      addPurchasedElectricityEmissions:
        product?.addPurchasedElectricityEmissions,
      addPurchasedHeatEmissions: product?.addPurchasedHeatEmissions,
      subtractExportedElectricityEmissions:
        product?.subtractExportedElectricityEmissions,
      subtractExportedHeatEmissions: product?.subtractExportedHeatEmissions,
      subtractGeneratedElectricityEmissions:
        product?.subtractGeneratedElectricityEmissions,
      subtractGeneratedHeatEmissions: product?.subtractGeneratedHeatEmissions,
      addEmissionsFromEios: product?.addEmissionsFromEios
    });
  };

  const handleChange = (product: FormData) => {
    if (formData.productRowId === product.productRowId) onChange(product);
    else handleProductChange(product.productRowId);
  };

  return productIsPublished(formData, query) ? (
    <ObjectField {...props} onChange={handleChange} />
  ) : (
    <>
      <Alert variant="danger">
        <strong>Warning:</strong> This version of the Product or Service has
        been archived. Please remove it and select an appropriate replacement
        (it may have the same name)
      </Alert>
      <ObjectField {...props} disabled onChange={handleChange} />
    </>
  );
};

export default createFragmentContainer(ProductFieldComponent, {
  query: graphql`
    fragment ProductField_query on Query {
      allProducts {
        edges {
          node {
            rowId
            units
            productState
            requiresEmissionAllocation
            requiresProductAmount
            isEnergyProduct
            addPurchasedElectricityEmissions
            addPurchasedHeatEmissions
            subtractExportedElectricityEmissions
            subtractExportedHeatEmissions
            subtractGeneratedElectricityEmissions
            subtractGeneratedHeatEmissions
            addEmissionsFromEios
          }
        }
      }
    }
  `
});
