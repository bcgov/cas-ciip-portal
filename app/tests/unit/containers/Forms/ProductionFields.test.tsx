import React from 'react';
import {shallow} from 'enzyme';
import {ProductFieldComponent} from 'containers/Forms/ProductField';
import productionSchema from 'schema/data/prod/form_json/production.json';
import {JSONSchema6} from 'json-schema';
import {getDefaultRegistry} from 'react-jsonschema-form/lib/utils';
import FormArrayFieldTemplate from 'containers/Forms/FormArrayFieldTemplate';
import FormObjectFieldTemplate from 'containers/Forms/FormObjectFieldTemplate';
import FormFieldTemplate from 'containers/Forms/FormFieldTemplate';
import {ProductField_query} from '__generated__/ProductField_query.graphql';

describe('The ProductionFields Component', () => {
  const query: ProductField_query = {
    ' $refType': 'ProductField_query',
    allProducts: {
      edges: [
        {
          node: {
            rowId: 1,
            name: 'foo',
            units: 'bar',
            requiresEmissionAllocation: true
          }
        }
      ]
    }
  };

  const idSchema: any = {$id: 'product_0'};
  const props = {
    schema: productionSchema.schema.definitions.product as JSONSchema6,
    uiSchema: productionSchema.uiSchema,
    idSchema,
    autofocus: false,
    disabled: false,
    errorSchema: null,
    formContext: {},
    readonly: false,
    required: false,
    registry: {
      ...getDefaultRegistry(),
      ArrayFieldTemplate: FormArrayFieldTemplate,
      ObjectFieldTemplate: FormObjectFieldTemplate,
      FieldTemplate: FormFieldTemplate
    } as any,

    name: 'product',
    query,
    formData: {
      productRowId: 1,
      productUnits: 'bar',
      productEmissions: 123,
      quantity: 456,
      requiresEmissionAllocation: true
    },
    onChange: jest.fn(),
    onBlur: jest.fn()
  };

  it('should match the snapshot when rendering a product', () => {
    const wrapper = shallow(<ProductFieldComponent {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
