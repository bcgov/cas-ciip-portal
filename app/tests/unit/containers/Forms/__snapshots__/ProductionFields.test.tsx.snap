// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ProductionFields Component should match the snapshot when rendering a product without additional data 1`] = `
<Fragment>
  <FormFieldTemplate
    classNames="form-group field field-string"
    formContext={Object {}}
    hidden={false}
    id="product.product"
    label="Product or Service"
    required={true}
    schema={
      Object {
        "enum": Array [
          "foo",
          "simpleFoo",
        ],
        "title": "Product or Service",
        "type": "string",
      }
    }
    uiSchema={Object {}}
  >
    <StringField
      autofocus={false}
      disabled={false}
      errorSchema={null}
      formContext={Object {}}
      formData="simpleFoo"
      idSchema={null}
      name="product"
      onChange={[Function]}
      readonly={false}
      registry={
        Object {
          "ArrayFieldTemplate": [Function],
          "FieldTemplate": [Function],
          "ObjectFieldTemplate": [Function],
          "definitions": Object {},
          "fields": Object {
            "AnyOfField": [Function],
            "ArrayField": [Function],
            "BooleanField": [Function],
            "DescriptionField": [Function],
            "NullField": [Function],
            "NumberField": [Function],
            "ObjectField": [Function],
            "OneOfField": [Function],
            "SchemaField": [Function],
            "StringField": [Function],
            "TitleField": [Function],
            "UnsupportedField": [Function],
          },
          "formContext": Object {},
          "widgets": Object {
            "AltDateTimeWidget": [Function],
            "AltDateWidget": [Function],
            "BaseInput": [Function],
            "CheckboxWidget": [Function],
            "CheckboxesWidget": [Function],
            "ColorWidget": [Function],
            "DateTimeWidget": [Function],
            "DateWidget": [Function],
            "EmailWidget": [Function],
            "FileWidget": [Function],
            "HiddenWidget": [Function],
            "PasswordWidget": [Function],
            "RadioWidget": [Function],
            "RangeWidget": [Function],
            "SelectWidget": [Function],
            "TextWidget": [Function],
            "TextareaWidget": [Function],
            "URLWidget": [Function],
            "UpDownWidget": [Function],
          },
        }
      }
      required={true}
      schema={
        Object {
          "enum": Array [
            "foo",
            "simpleFoo",
          ],
          "title": "Product or Service",
          "type": "string",
        }
      }
      uiSchema={Object {}}
    />
  </FormFieldTemplate>
  <FormFieldTemplate
    classNames="form-group field field-number"
    formContext={Object {}}
    hidden={false}
    id="product.allocationFactor"
    label="Product allocation factor"
    required={true}
    schema={
      Object {
        "title": "Product allocation factor",
        "type": "number",
      }
    }
    uiSchema={Object {}}
  >
    <NumberField
      autofocus={false}
      disabled={false}
      errorSchema={null}
      formContext={Object {}}
      idSchema={null}
      name="allocationFactor"
      onChange={[Function]}
      readonly={false}
      registry={
        Object {
          "ArrayFieldTemplate": [Function],
          "FieldTemplate": [Function],
          "ObjectFieldTemplate": [Function],
          "definitions": Object {},
          "fields": Object {
            "AnyOfField": [Function],
            "ArrayField": [Function],
            "BooleanField": [Function],
            "DescriptionField": [Function],
            "NullField": [Function],
            "NumberField": [Function],
            "ObjectField": [Function],
            "OneOfField": [Function],
            "SchemaField": [Function],
            "StringField": [Function],
            "TitleField": [Function],
            "UnsupportedField": [Function],
          },
          "formContext": Object {},
          "widgets": Object {
            "AltDateTimeWidget": [Function],
            "AltDateWidget": [Function],
            "BaseInput": [Function],
            "CheckboxWidget": [Function],
            "CheckboxesWidget": [Function],
            "ColorWidget": [Function],
            "DateTimeWidget": [Function],
            "DateWidget": [Function],
            "EmailWidget": [Function],
            "FileWidget": [Function],
            "HiddenWidget": [Function],
            "PasswordWidget": [Function],
            "RadioWidget": [Function],
            "RangeWidget": [Function],
            "SelectWidget": [Function],
            "TextWidget": [Function],
            "TextareaWidget": [Function],
            "URLWidget": [Function],
            "UpDownWidget": [Function],
          },
        }
      }
      required={true}
      schema={
        Object {
          "title": "Product allocation factor",
          "type": "number",
        }
      }
      uiSchema={Object {}}
    />
  </FormFieldTemplate>
  <FormFieldTemplate
    classNames="form-group field field-number"
    formContext={Object {}}
    hidden={false}
    id="product.quantity"
    label="Quantity"
    required={true}
    schema={
      Object {
        "title": "Quantity",
        "type": "number",
      }
    }
    uiSchema={Object {}}
  >
    <NumberField
      autofocus={false}
      disabled={false}
      errorSchema={null}
      formContext={Object {}}
      idSchema={null}
      name="quantity"
      onChange={[Function]}
      readonly={false}
      registry={
        Object {
          "ArrayFieldTemplate": [Function],
          "FieldTemplate": [Function],
          "ObjectFieldTemplate": [Function],
          "definitions": Object {},
          "fields": Object {
            "AnyOfField": [Function],
            "ArrayField": [Function],
            "BooleanField": [Function],
            "DescriptionField": [Function],
            "NullField": [Function],
            "NumberField": [Function],
            "ObjectField": [Function],
            "OneOfField": [Function],
            "SchemaField": [Function],
            "StringField": [Function],
            "TitleField": [Function],
            "UnsupportedField": [Function],
          },
          "formContext": Object {},
          "widgets": Object {
            "AltDateTimeWidget": [Function],
            "AltDateWidget": [Function],
            "BaseInput": [Function],
            "CheckboxWidget": [Function],
            "CheckboxesWidget": [Function],
            "ColorWidget": [Function],
            "DateTimeWidget": [Function],
            "DateWidget": [Function],
            "EmailWidget": [Function],
            "FileWidget": [Function],
            "HiddenWidget": [Function],
            "PasswordWidget": [Function],
            "RadioWidget": [Function],
            "RangeWidget": [Function],
            "SelectWidget": [Function],
            "TextWidget": [Function],
            "TextareaWidget": [Function],
            "URLWidget": [Function],
            "UpDownWidget": [Function],
          },
        }
      }
      required={true}
      schema={
        Object {
          "title": "Quantity",
          "type": "number",
        }
      }
      uiSchema={Object {}}
    />
  </FormFieldTemplate>
  <FormFieldTemplate
    classNames="form-group field field-string"
    formContext={Object {}}
    hidden={false}
    id="product.productUnits"
    label="Units"
    required={false}
    schema={
      Object {
        "title": "Units",
        "type": "string",
      }
    }
    uiSchema={Object {}}
  >
    <StringField
      autofocus={false}
      disabled={true}
      errorSchema={null}
      formContext={Object {}}
      formData="baz"
      idSchema={null}
      name="units"
      onChange={[Function]}
      readonly={true}
      registry={
        Object {
          "ArrayFieldTemplate": [Function],
          "FieldTemplate": [Function],
          "ObjectFieldTemplate": [Function],
          "definitions": Object {},
          "fields": Object {
            "AnyOfField": [Function],
            "ArrayField": [Function],
            "BooleanField": [Function],
            "DescriptionField": [Function],
            "NullField": [Function],
            "NumberField": [Function],
            "ObjectField": [Function],
            "OneOfField": [Function],
            "SchemaField": [Function],
            "StringField": [Function],
            "TitleField": [Function],
            "UnsupportedField": [Function],
          },
          "formContext": Object {},
          "widgets": Object {
            "AltDateTimeWidget": [Function],
            "AltDateWidget": [Function],
            "BaseInput": [Function],
            "CheckboxWidget": [Function],
            "CheckboxesWidget": [Function],
            "ColorWidget": [Function],
            "DateTimeWidget": [Function],
            "DateWidget": [Function],
            "EmailWidget": [Function],
            "FileWidget": [Function],
            "HiddenWidget": [Function],
            "PasswordWidget": [Function],
            "RadioWidget": [Function],
            "RangeWidget": [Function],
            "SelectWidget": [Function],
            "TextWidget": [Function],
            "TextareaWidget": [Function],
            "URLWidget": [Function],
            "UpDownWidget": [Function],
          },
        }
      }
      required={false}
      schema={
        Object {
          "title": "Units",
          "type": "string",
        }
      }
      uiSchema={Object {}}
    />
  </FormFieldTemplate>
</Fragment>
`;

exports[`The ProductionFields Component should match the snapshot when rendering additional data 1`] = `
<Fragment>
  <FormFieldTemplate
    classNames="form-group field field-string"
    formContext={Object {}}
    hidden={false}
    id="product.product"
    label="Product or Service"
    required={true}
    schema={
      Object {
        "enum": Array [
          "foo",
          "simpleFoo",
        ],
        "title": "Product or Service",
        "type": "string",
      }
    }
    uiSchema={Object {}}
  >
    <StringField
      autofocus={false}
      disabled={false}
      errorSchema={null}
      formContext={Object {}}
      formData="foo"
      idSchema={null}
      name="product"
      onChange={[Function]}
      readonly={false}
      registry={
        Object {
          "ArrayFieldTemplate": [Function],
          "FieldTemplate": [Function],
          "ObjectFieldTemplate": [Function],
          "definitions": Object {},
          "fields": Object {
            "AnyOfField": [Function],
            "ArrayField": [Function],
            "BooleanField": [Function],
            "DescriptionField": [Function],
            "NullField": [Function],
            "NumberField": [Function],
            "ObjectField": [Function],
            "OneOfField": [Function],
            "SchemaField": [Function],
            "StringField": [Function],
            "TitleField": [Function],
            "UnsupportedField": [Function],
          },
          "formContext": Object {},
          "widgets": Object {
            "AltDateTimeWidget": [Function],
            "AltDateWidget": [Function],
            "BaseInput": [Function],
            "CheckboxWidget": [Function],
            "CheckboxesWidget": [Function],
            "ColorWidget": [Function],
            "DateTimeWidget": [Function],
            "DateWidget": [Function],
            "EmailWidget": [Function],
            "FileWidget": [Function],
            "HiddenWidget": [Function],
            "PasswordWidget": [Function],
            "RadioWidget": [Function],
            "RangeWidget": [Function],
            "SelectWidget": [Function],
            "TextWidget": [Function],
            "TextareaWidget": [Function],
            "URLWidget": [Function],
            "UpDownWidget": [Function],
          },
        }
      }
      required={true}
      schema={
        Object {
          "enum": Array [
            "foo",
            "simpleFoo",
          ],
          "title": "Product or Service",
          "type": "string",
        }
      }
      uiSchema={Object {}}
    />
  </FormFieldTemplate>
  <FormFieldTemplate
    classNames="form-group field field-number"
    formContext={Object {}}
    hidden={false}
    id="product.allocationFactor"
    label="Product allocation factor"
    required={true}
    schema={
      Object {
        "title": "Product allocation factor",
        "type": "number",
      }
    }
    uiSchema={Object {}}
  >
    <NumberField
      autofocus={false}
      disabled={false}
      errorSchema={null}
      formContext={Object {}}
      idSchema={null}
      name="allocationFactor"
      onChange={[Function]}
      readonly={false}
      registry={
        Object {
          "ArrayFieldTemplate": [Function],
          "FieldTemplate": [Function],
          "ObjectFieldTemplate": [Function],
          "definitions": Object {},
          "fields": Object {
            "AnyOfField": [Function],
            "ArrayField": [Function],
            "BooleanField": [Function],
            "DescriptionField": [Function],
            "NullField": [Function],
            "NumberField": [Function],
            "ObjectField": [Function],
            "OneOfField": [Function],
            "SchemaField": [Function],
            "StringField": [Function],
            "TitleField": [Function],
            "UnsupportedField": [Function],
          },
          "formContext": Object {},
          "widgets": Object {
            "AltDateTimeWidget": [Function],
            "AltDateWidget": [Function],
            "BaseInput": [Function],
            "CheckboxWidget": [Function],
            "CheckboxesWidget": [Function],
            "ColorWidget": [Function],
            "DateTimeWidget": [Function],
            "DateWidget": [Function],
            "EmailWidget": [Function],
            "FileWidget": [Function],
            "HiddenWidget": [Function],
            "PasswordWidget": [Function],
            "RadioWidget": [Function],
            "RangeWidget": [Function],
            "SelectWidget": [Function],
            "TextWidget": [Function],
            "TextareaWidget": [Function],
            "URLWidget": [Function],
            "UpDownWidget": [Function],
          },
        }
      }
      required={true}
      schema={
        Object {
          "title": "Product allocation factor",
          "type": "number",
        }
      }
      uiSchema={Object {}}
    />
  </FormFieldTemplate>
  <ObjectField
    autofocus={false}
    disabled={false}
    errorSchema={null}
    formContext={Object {}}
    formData={
      Object {
        "additionalFoo": "foofoo",
      }
    }
    idSchema={null}
    name={null}
    onChange={[Function]}
    readonly={false}
    registry={
      Object {
        "ArrayFieldTemplate": [Function],
        "FieldTemplate": [Function],
        "ObjectFieldTemplate": [Function],
        "definitions": Object {},
        "fields": Object {
          "AnyOfField": [Function],
          "ArrayField": [Function],
          "BooleanField": [Function],
          "DescriptionField": [Function],
          "NullField": [Function],
          "NumberField": [Function],
          "ObjectField": [Function],
          "OneOfField": [Function],
          "SchemaField": [Function],
          "StringField": [Function],
          "TitleField": [Function],
          "UnsupportedField": [Function],
        },
        "formContext": Object {},
        "widgets": Object {
          "AltDateTimeWidget": [Function],
          "AltDateWidget": [Function],
          "BaseInput": [Function],
          "CheckboxWidget": [Function],
          "CheckboxesWidget": [Function],
          "ColorWidget": [Function],
          "DateTimeWidget": [Function],
          "DateWidget": [Function],
          "EmailWidget": [Function],
          "FileWidget": [Function],
          "HiddenWidget": [Function],
          "PasswordWidget": [Function],
          "RadioWidget": [Function],
          "RangeWidget": [Function],
          "SelectWidget": [Function],
          "TextWidget": [Function],
          "TextareaWidget": [Function],
          "URLWidget": [Function],
          "UpDownWidget": [Function],
        },
      }
    }
    required={true}
    schema={
      Object {
        "properties": Object {
          "additionalFoo": Object {
            "type": "string",
          },
        },
        "type": "object",
      }
    }
    uiSchema={
      Object {
        "additionalFoo": Object {
          "ui:col-md": 12,
        },
      }
    }
  />
</Fragment>
`;
